cli/src/common/pcluster_api.py

```diff
@@ -0,0 +1,165 @@
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import logging
+
+from common.aws.aws_api import AWSApi
+from pcluster.models.cluster import ClusterActionError, ClusterStack, Tag
+from pcluster.schemas.cluster_schema import ClusterSchema
+from pcluster.templates.cdk_builder import CDKTemplateBuilder
+from pcluster.utils import get_installed_version, get_stack_name
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ApiFailure:
+    """Represent a generic api error."""
+
+    def __init__(self, message: str = None, validation_failures: list = None):
+        self.message = message or "Something went wrong."
+        self.validation_failures = validation_failures or []
+
+
+class ClusterInfo:
+    """Minimal representation of a running cluster."""
+
+    def __init__(self, stack: ClusterStack, cluster_name: str):
+        self.stack_status = stack.status
+        self.stack_arn = stack.id
+        self.id = cluster_name
+        self.region = stack.region
+        self.version = stack.version
+        self.status = stack.status
+
+    def __repr__(self):
+        return json.dumps(self.__dict__)
+
+
+class PclusterApi:
+    """Proxy class for all Pcluster API commands used in the CLI."""
+
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def create_cluster(cluster_config: dict, cluster_name: str, region: str = None, disable_rollback: bool = False):
+        """
+        Load cluster model from cluster_config and create stack.
+
+        :param cluster_config: cluster configuration (yaml dict)
+        :param cluster_name: the name to assign to the cluster
+        :param region: AWS region
+        :param disable_rollback: Disable rollback in case of failures
+        """
+        try:
+            # check cluster existence
+            stack_name = get_stack_name(cluster_name)
+            if AWSApi.instance().cfn.stack_exists(stack_name):
+                return ApiFailure(f"Cluster {cluster_name} already exists.")
+
+            # Generate model from config dict and validate
+            cluster = _init_cluster_from_config(cluster_name, cluster_config)
+            validation_failures = cluster.validate()
+
+            if validation_failures:
+                # TODO skip validation errors
+                return ApiFailure(message="Configuration is invalid", validation_failures=validation_failures)
+            else:
+                # Add tags information to the cluster
+                version = get_installed_version()
+                tags = cluster.tags or []
+                tags.append(Tag(key="Version", value=version))
+                tags = [{"Key": tag.key, "Value": tag.value} for tag in tags]
```

Comment:
Why append the default tag here, instead of doing it in model layer?

Reply:
It's a good point. 
I don't think that the version tag should be added when parsing the config because if you are running version 3.1 and you parse a configuration of an existing cluster (created with version 3.0) you'll add 3.1 in the config and it's wrong.
I think this particular tag should be added only when creating the cluster, we only need it at cluster creation time, but let's talk about it. This can be interesting for @yuleiwan too.

Reply:
Very true, the default tag should be here